commcare_connect/form_receiver/processor.py

```diff
@@ -272,83 +276,99 @@ def process_deliver_unit(user, xform: XForm, app: CommCareApp, opportunity: Oppo
+            .annotate(rank=Window(Rank(), order_by="date_created"))
+            .filter(rank=counts["entity"] + 1)
+            .order_by("rank")
+            .last()
         )
-        completed_work_needs_save = False
-        today = datetime.date.today()
-        paymentunit_startdate = payment_unit.start_date if payment_unit else None
-        if opportunity.start_date > today or (paymentunit_startdate and paymentunit_startdate > today):
-            completed_work = None
-            user_visit.status = VisitValidationStatus.trial
-        else:
-            completed_work, _ = CompletedWork.objects.get_or_create(
-                opportunity_access=access,
-                entity_id=entity_id,
-                payment_unit=payment_unit,
-                defaults={"entity_name": entity_name},
+        if completed_work is None:
+            completed_work = CompletedWork(
+                opportunity_access=access, entity_id=entity_id, payment_unit=payment_unit, entity_name=entity_name
             )
-            user_visit.completed_work = completed_work
-            if (
-                counts["daily"] >= payment_unit.max_daily
-                or counts["total"] >= claim_limit.max_visits
-                or (today > claim.end_date or (claim_limit.end_date and today > claim_limit.end_date))
-            ):
-                user_visit.status = VisitValidationStatus.over_limit
-                if not completed_work.status == CompletedWorkStatus.over_limit:
-                    completed_work.status = CompletedWorkStatus.over_limit
-                    completed_work_needs_save = True
-            elif counts["entity"] > 0:
-                user_visit.status = VisitValidationStatus.duplicate
-        flags = clean_form_submission(access, user_visit, xform)
-        if access.suspended:
-            flags.append(["user_suspended", "This user is suspended from the opportunity."])
-            user_visit.status = VisitValidationStatus.rejected
-            completed_work.status = CompletedWorkStatus.rejected
-        if flags:
-            user_visit.flagged = True
-            user_visit.flag_reason = {"flags": flags}
-
-        if (
-            opportunity.auto_approve_visits
-            and user_visit.status == VisitValidationStatus.pending
-            and not user_visit.flagged
-        ):
-            user_visit.status = VisitValidationStatus.approved
-            user_visit.review_status = VisitReviewStatus.agree
+            completed_work.save()
 
+    try:
+        user_visit.completed_work = completed_work
+        user_visit.save()
+    except IntegrityError as e:
+        if UNIQUE_USER_VISIT_CONSTRAINT not in str(e):
+            raise e
+        completed_work = CompletedWork(
+            opportunity_access=access, entity_id=entity_id, payment_unit=payment_unit, entity_name=entity_name
+        )
+        completed_work.save()
```

Comment:
Does't this end up creating two completed_work objects, first one unlinked ([above](https://github.com/dimagi/commcare-connect/pull/678/files#diff-0d916b95ee38daa9788eca7b8b3266f406c5c791cab6ff6bda1d5aeefc84d5d2R316)), second one linked here?